crates/index-scheduler/src/scheduler/process_snapshot_creation.rs

```diff
@@ -206,4 +230,340 @@ impl IndexScheduler {
+            // Note for the future: use with_added_extension, it's prettier
+            let object_path = s3_snapshot_prefix.join(format!("{db_name}.snapshot"));
+            let object = object_path.to_slash().expect("Invalid UTF-8 path").into_owned();
+
+            let action = bucket.create_multipart_upload(Some(&credential), &object);
+            let url = action.sign(s3_signature_duration);
+            let client = Client::new();
+            let resp = client.post(url).send().await.map_err(Error::S3HttpError)?;
+            let status = resp.status();
+            let body = match resp.error_for_status_ref() {
+                Ok(_) => resp.text().await.map_err(Error::S3HttpError)?,
+                Err(_) => {
+                    return Err(Error::S3Error {
+                        status,
+                        body: resp.text().await.unwrap_or_default(),
+                    })
+                }
+            };
+
+            let multipart = CreateMultipartUpload::parse_response(&body)
+                .map_err(|e| Error::S3XmlError(Box::new(e)))?;
+            tracing::debug!("Starting the upload of the snapshot to {object}");
+
+            // We use this bumpalo for etags strings.
+            let bump = bumpalo::Bump::new();
+            let mut etags = Vec::<&str>::new();
+            let mut in_flight =
+                VecDeque::<(JoinHandle<reqwest::Result<Response>>, Bytes)>::with_capacity(
+                    s3_max_in_flight_parts.get(),
+                );
+
+            // Part numbers start at 1 and cannot be larger than 10k
+            for part_number in 1u16.. {
+                if must_stop_processing.get() {
+                    return Err(Error::AbortedTask);
+                }
+
+                let part_upload = bucket.upload_part(
+                    Some(&credential),
+                    &object,
+                    part_number,
+                    multipart.upload_id(),
+                );
+                let url = part_upload.sign(s3_signature_duration);
+
+                // Wait for a buffer to be ready if there are in-flight parts that landed
+                let mut buffer = if in_flight.len() >= s3_max_in_flight_parts.get() {
+                    let (request, buffer) = in_flight.pop_front().unwrap();
+                    let resp = request.await.unwrap().map_err(Error::S3HttpError)?;
+                    let resp = match resp.error_for_status_ref() {
+                        Ok(response) => response,
+                        Err(_) => {
+                            return Err(Error::S3Error {
+                                status: resp.status(),
+                                body: resp.text().await.unwrap_or_default(),
+                            });
+                        }
+                    };
+                    let etag =
+                        resp.headers().get(ETAG).expect("every UploadPart request returns an Etag");
```

Comment:
Since this depends of S3 rather than our code, I feel like this should be an Error rather than a panic